cmd/derper/bootstrap_dns.go

```diff
@@ -48,21 +87,42 @@ func refreshBootstrapDNS() {
 		}
 		dnsEntries[name] = addrs
 	}
-	j, err := json.MarshalIndent(dnsEntries, "", "\t")
-	if err != nil {
-		// leave the old values in place
-		return
-	}
-	dnsCache.Store(j)
+	return dnsEntries
 }
 
 func handleBootstrapDNS(w http.ResponseWriter, r *http.Request) {
 	bootstrapDNSRequests.Add(1)
+
 	w.Header().Set("Content-Type", "application/json")
-	j := dnsCache.Load()
-	// Bootstrap DNS requests occur cross-regions,
-	// and are randomized per request,
-	// so keeping a connection open is pointlessly expensive.
+	// Bootstrap DNS requests occur cross-regions, and are randomized per
+	// request, so keeping a connection open is pointlessly expensive.
 	w.Header().Set("Connection", "close")
+
+	// Try answering a query from our hidden map first
+	if q := r.URL.Query().Get("q"); q != "" {
+		if ips, ok := unpublishedDNSCache.Load()[q]; ok {
+			unpublishedDNSHits.Add(1)
+
+			// Only return the specific query, not everything.
+			m := dnsEntryMap{q: ips}
+			j, err := json.MarshalIndent(m, "", "\t")
+			if err == nil {
+				w.Write(j)
+				return
+			}
+		} else {
+			unpublishedDNSMisses.Add(1)
+		}
+
+		// Also track whether this is a cache hit in our published DNS cache.
+		if _, ok := dnsCache.Load()[q]; ok {
+			publishedDNSHits.Add(1)
+		} else {
+			publishedDNSMisses.Add(1)
+		}
```

Comment:
I was imagining a different definition for these counters:

* if the "q" name is one of the 4 at https://derp10.tailscale.com/bootstrap-dns and we have a non-empty value, that's a `publishedDNSHits`.
* else, if the "q" name is one of the 4 and its value is currently `[]`, that's a `publishedDNSMisses`
* else, ... unpublished hits vs miss

But a given `q` can't be both published and unpublished.

Reply:
Makes sense; changed that